MaterialLib/MPL/Properties/ExponentialProperty.cpp

```diff
@@ -0,0 +1,69 @@
+/**
+ * \file
+ *
+ * \copyright
+ * Copyright (c) 2012-2019, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <cmath>
+
+#include "MaterialLib/MPL/Properties/ExponentialProperty.h"
+
+namespace MaterialPropertyLib
+{
+ExponentialProperty::ExponentialProperty(
+    PropertyDataType const& property_reference_value,
+    ExponentData const& v)
+    : _exponent_data(v)
+{
+    _value = property_reference_value;
+}
+
+PropertyDataType ExponentialProperty::value(
+    VariableArray const& variable_array) const
+{
+    return boost::get<double>(_value) *
+           std::exp(
+               -boost::get<double>(_exponent_data.factor) *
+               (boost::get<double>(
+                    variable_array[static_cast<int>(_exponent_data.type)]) -
+                boost::get<double>(_exponent_data.reference_condition)));
+}
+
+PropertyDataType ExponentialProperty::dValue(
+    VariableArray const& variable_array, Variable const primary_variable) const
+{
+    return _exponent_data.type == primary_variable
+               ? -boost::get<double>(_value) *
+                     boost::get<double>(_exponent_data.factor) *
+                     std::exp(
+                         -boost::get<double>(_exponent_data.factor) *
+                         (boost::get<double>(variable_array[static_cast<int>(
+                              _exponent_data.type)]) -
+                          boost::get<double>(
+                              _exponent_data.reference_condition)))
+               : decltype(_value){};
+}
+
+PropertyDataType ExponentialProperty::d2Value(
+    VariableArray const& variable_array,
+    Variable const pv1,
+    Variable const pv2) const
+{
+    return _exponent_data.type == pv1 && _exponent_data.type == pv2
+               ? boost::get<double>(_value) *
+                     std::pow(boost::get<double>(_exponent_data.factor), 2) *
```

Comment:
Don't use `std::pow` for integer powers---it's slow. Use `boost::math::pow<2>(x)` in this case.

Reply:
Did not know that the integer overloads of `std::pow` are gone and the ints were casted to floating point values. Will change it as suggested.

Reply:
I remember the optimization when I was refactoring Ehlers material model functions. Using templated boost version was much faster.
Now I tested some simple code and for std::pow(x,2) the assembly is the same as for x*x, but not for std::pow(x, 3) and x*x*x. For 3rd power the std::pow is ca. 2 times slower. http://quick-bench.com/smUihBRSJUMUXgBwKhEP_pDE14w

The overloads are there, but I'm not sure, why it is inlined in one case, but not in another.